File: src/middleware/security.js
import crypto from 'crypto';

// ---------------------------------------------------------------------------
// 1. SECURITY RESPONSE HEADERS
//    Call applySecurityHeaders(app) once in server.js before any routes.
// ---------------------------------------------------------------------------
export function applySecurityHeaders(app) {
  app.use((req, res, next) => {
    res.setHeader('X-Content-Type-Options', 'nosniff');
    res.setHeader('X-Frame-Options', 'DENY');
    res.setHeader('X-XSS-Protection', '1; mode=block');
    res.setHeader('Referrer-Policy', 'no-referrer');
    res.setHeader(
      'Permissions-Policy',
      'geolocation=(), camera=(), microphone=()'
    );
    // Remove fingerprinting header
    res.removeHeader('X-Powered-By');
    next();
  });
}

// ---------------------------------------------------------------------------
// 2. ADMIN API KEY VERIFICATION (timing-safe)
//    Replaces the inline verifyAdminApiKey in admin.js and razorpay.js.
//    Hard-fails if ADMIN_API_KEY env var is not configured — no fallback.
// ---------------------------------------------------------------------------
export function requireAdminKey(req, res, next) {
  const configuredKey = process.env.ADMIN_API_KEY;

  if (!configuredKey) {
    console.error(
      '[SECURITY] ADMIN_API_KEY is not set in environment. Refusing all admin requests.'
    );
    return res.status(503).json({ error: 'Server misconfiguration. Contact administrator.' });
  }

  const providedKey = req.headers['x-admin-key'] || '';

  // Timing-safe comparison prevents brute-force timing attacks
  let keysMatch = false;
  try {
    const a = Buffer.from(configuredKey, 'utf8');
    const b = Buffer.from(providedKey, 'utf8');
    if (a.length === b.length) {
      keysMatch = crypto.timingSafeEqual(a, b);
    }
  } catch {
    keysMatch = false;
  }

  if (!keysMatch) {
    return res.status(401).json({ error: 'Unauthorized: Invalid or missing Admin API Key' });
  }

  next();
}

// ---------------------------------------------------------------------------
// 3. WHATSAPP WEBHOOK SIGNATURE VERIFICATION
//    Meta sends an X-Hub-Signature-256 header on every POST.
//    We verify it using your Meta App Secret (WHATSAPP_APP_SECRET).
//
//    IMPORTANT: This middleware must run BEFORE express.json() parses the body,
//    so it relies on the raw Buffer stored in req.rawBody by server.js.
// ---------------------------------------------------------------------------
export function verifyWhatsAppSignature(req, res, next) {
  const appSecret = process.env.WHATSAPP_APP_SECRET;

  // If secret is not configured (e.g. local dev without Meta credentials),
  // skip signature check but log a warning.
  if (!appSecret) {
    console.warn(
      '[SECURITY WARNING] WHATSAPP_APP_SECRET not set — skipping webhook signature verification. ' +
      'Set this in production!'
    );
    return next();
  }

  const sigHeader = req.headers['x-hub-signature-256'];
  if (!sigHeader) {
    console.warn('[SECURITY] WhatsApp webhook received with no signature header — rejected.');
    return res.status(403).json({ error: 'Forbidden: Missing webhook signature' });
  }

  const rawBody = req.rawBody; // set by the rawBodyCapture middleware in server.js
  if (!rawBody) {
    console.error('[SECURITY] rawBody not available for signature check.');
    return res.status(400).json({ error: 'Bad Request' });
  }

  const expectedSig =
    'sha256=' +
    crypto.createHmac('sha256', appSecret).update(rawBody).digest('hex');

  let sigMatch = false;
  try {
    const a = Buffer.from(expectedSig, 'utf8');
    const b = Buffer.from(sigHeader, 'utf8');
    if (a.length === b.length) {
      sigMatch = crypto.timingSafeEqual(a, b);
    }
  } catch {
    sigMatch = false;
  }

  if (!sigMatch) {
    console.warn('[SECURITY] WhatsApp webhook signature mismatch — request rejected.');
    return res.status(403).json({ error: 'Forbidden: Invalid webhook signature' });
  }

  next();
}

// ---------------------------------------------------------------------------
// 4. TELEGRAM WEBHOOK SECRET TOKEN VERIFICATION
//    Telegram sends X-Telegram-Bot-Api-Secret-Token when you set a secret
//    token via setWebhook. Configure TELEGRAM_WEBHOOK_SECRET in your .env.
// ---------------------------------------------------------------------------
export function verifyTelegramToken(req, res, next) {
  const expectedToken = process.env.TELEGRAM_WEBHOOK_SECRET;

  // If not configured, skip (warn in dev, enforce in production)
  if (!expectedToken) {
    if (process.env.NODE_ENV === 'production') {
      console.error(
        '[SECURITY] TELEGRAM_WEBHOOK_SECRET not set in production — rejecting Telegram webhook.'
      );
      return res.status(403).json({ error: 'Forbidden: Telegram webhook not secured' });
    }
    console.warn(
      '[SECURITY WARNING] TELEGRAM_WEBHOOK_SECRET not set — skipping Telegram webhook verification.'
    );
    return next();
  }

  const providedToken = req.headers['x-telegram-bot-api-secret-token'] || '';

  let tokensMatch = false;
  try {
    const a = Buffer.from(expectedToken, 'utf8');
    const b = Buffer.from(providedToken, 'utf8');
    if (a.length === b.length) {
      tokensMatch = crypto.timingSafeEqual(a, b);
    }
  } catch {
    tokensMatch = false;
  }

  if (!tokensMatch) {
    console.warn('[SECURITY] Telegram webhook token mismatch — request rejected.');
    return res.status(403).json({ error: 'Forbidden: Invalid Telegram webhook token' });
  }

  next();
}

// ---------------------------------------------------------------------------
// 5. REPORT FILE AUTHENTICATION
//    Protects express.static '/reports' directory — requires admin key.
// ---------------------------------------------------------------------------
export function requireAdminKeyForReports(req, res, next) {
  const configuredKey = process.env.ADMIN_API_KEY;
  const providedKey =
    req.headers['x-admin-key'] ||
    req.query['admin_key'] || // allow ?admin_key= in URL for direct browser access
    '';

  if (!configuredKey) {
    return res.status(503).json({ error: 'Server misconfiguration.' });
  }

  let keysMatch = false;
  try {
    const a = Buffer.from(configuredKey, 'utf8');
    const b = Buffer.from(providedKey, 'utf8');
    if (a.length === b.length) {
      keysMatch = crypto.timingSafeEqual(a, b);
    }
  } catch {
    keysMatch = false;
  }

  if (!keysMatch) {
    return res.status(401).json({ error: 'Unauthorized: Admin key required to access reports' });
  }

  next();
}

// ---------------------------------------------------------------------------
// 6. INPUT SANITIZER
//    Trims whitespace and caps string length to prevent oversized DB writes.
//    Call: sanitizeInput(str, 500)
// ---------------------------------------------------------------------------
export function sanitizeInput(str, maxLength = 500) {
  if (typeof str !== 'string') return '';
  return str.trim().slice(0, maxLength);
}

File: strikit-python/app/middleware/security.py
"""
Security middleware — HMAC signature verification, admin key auth, input sanitization.
Timing-safe comparisons throughout to prevent brute-force timing attacks.
"""
import hmac
import hashlib
import logging
from fastapi import Request, HTTPException, Depends
from fastapi.security import APIKeyHeader
from app.config import settings

logger = logging.getLogger(__name__)

admin_key_header = APIKeyHeader(name="x-admin-key", auto_error=False)


# ── Admin API Key Verification (timing-safe) ──

async def require_admin_key(api_key: str = Depends(admin_key_header)):
    """FastAPI dependency — verify admin API key with timing-safe comparison."""
    configured_key = settings.ADMIN_API_KEY
    if not configured_key:
        logger.error("[SECURITY] ADMIN_API_KEY not set. Refusing all admin requests.")
        raise HTTPException(status_code=503, detail="Server misconfiguration. Contact administrator.")

    if not api_key or not hmac.compare_digest(configured_key, api_key):
        raise HTTPException(status_code=401, detail="Unauthorized: Invalid or missing Admin API Key")


# ── WhatsApp Webhook Signature Verification ──

async def verify_whatsapp_signature(request: Request):
    """Verify Meta X-Hub-Signature-256 HMAC signature."""
    app_secret = settings.WHATSAPP_APP_SECRET
    if not app_secret:
        logger.warning("[SECURITY] WHATSAPP_APP_SECRET not set — skipping signature verification.")
        return

    sig_header = request.headers.get("x-hub-signature-256", "")
    if not sig_header:
        logger.warning("[SECURITY] WhatsApp webhook: missing signature header — rejected.")
        raise HTTPException(status_code=403, detail="Forbidden: Missing webhook signature")

    raw_body = await request.body()
    expected_sig = "sha256=" + hmac.new(
        app_secret.encode(), raw_body, hashlib.sha256
    ).hexdigest()

    if not hmac.compare_digest(expected_sig, sig_header):
        logger.warning("[SECURITY] WhatsApp webhook: signature mismatch — rejected.")
        raise HTTPException(status_code=403, detail="Forbidden: Invalid webhook signature")


# ── Razorpay Webhook Signature Verification ──

def verify_razorpay_signature(raw_body: bytes, signature: str) -> bool:
    """Verify Razorpay webhook HMAC-SHA256 signature."""
    secret = settings.RAZORPAY_WEBHOOK_SECRET
    if not secret:
        logger.warning("[SECURITY] RAZORPAY_WEBHOOK_SECRET not set — skipping verification.")
        return True

    expected = hmac.new(secret.encode(), raw_body, hashlib.sha256).hexdigest()
    return hmac.compare_digest(expected, signature)


# ── Telegram Webhook Secret Token Verification ──

async def verify_telegram_token(request: Request):
    """Verify Telegram X-Telegram-Bot-Api-Secret-Token header."""
    expected_token = settings.TELEGRAM_WEBHOOK_SECRET
    if not expected_token:
        if settings.is_production:
            logger.error("[SECURITY] TELEGRAM_WEBHOOK_SECRET not set in production — rejecting.")
            raise HTTPException(status_code=403, detail="Forbidden: Telegram webhook not secured")
        logger.warning("[SECURITY] TELEGRAM_WEBHOOK_SECRET not set — skipping verification.")
        return

    provided_token = request.headers.get("x-telegram-bot-api-secret-token", "")
    if not hmac.compare_digest(expected_token, provided_token):
        logger.warning("[SECURITY] Telegram webhook: token mismatch — rejected.")
        raise HTTPException(status_code=403, detail="Forbidden: Invalid Telegram webhook token")


# ── Input Sanitizer ──

def sanitize_input(text: str, max_length: int = 500) -> str:
    """Trim whitespace and cap string length to prevent oversized DB writes."""
    if not isinstance(text, str):
        return ""
    return text.strip()[:max_length]
